src/descriptor/satisfied_contraints.rs

```diff
@@ -0,0 +1,1482 @@
+                        Some(Err(Error::CouldnotEvaluate(
+                            "PK evaluation error: Signature not present".to_owned())))
+                    }
+                }
+                Err(..) => Some(Err(
+                    Error::CouldnotEvaluate("Could not parse pubkey".to_owned())))
+            }
+        } else {
+            Some(Err(Error::CouldnotEvaluate("Pkh evaluation failed".to_owned())))
+        }
+    }
+
+    /// Helper function to evaluate a After and AfterF Node. Takes no arugment from stack
+    /// `n CHECKSEQUENCEVERIFY 0NOTEQUAL` and `n CHECKSEQUENCEVERIFY`
+    /// Ideally this should return int value as n: build_scriptint(t as i64)),
+    /// But since this is a T value and it will never be popped, it is safe to
+    /// push Satisfied.
+    fn evaluate_after<'desc, Pk, Pkh>(
+        &mut self,
+        n: &'desc u32,
+        age: &'desc u32,
+    ) -> Option<Result<SatisfiedConstraint<'desc, 'stack, Pk, Pkh>, Error>>
+        where Pk: ToPublicKey, Pkh: ToPublicKeyHash
+    {
+        if *age >= *n {
+            self.push(StackElement::Satisfied);
+            Some(Ok(SatisfiedConstraint::RelativeTimeLock { time: n }))
+        } else {
+            Some(Err(Error::RelativeLocktimeNotMet(n.clone())))
+        }
+    }
+
+    /// Helper function to evaluate a Older and OlderF Node. Takes no arugment from stack
+    /// `n CHECKLOCKTIMEVERIFY 0NOTEQUAL` and `n CHECKLOCKTIMEVERIFY`
+    /// Ideally this should return int value as n: build_scriptint(t as i64)),
+    /// But since this is a T value and it will never be popped, it is safe to
+    /// push Satisfied instead of n.
+    fn evaluate_older<'desc, Pk, Pkh>(
+        &mut self,
+        n: &'desc u32,
+        height: &'desc u32,
+    ) -> Option<Result<SatisfiedConstraint<'desc, 'stack, Pk, Pkh>, Error>>
+        where Pk: ToPublicKey, Pkh: ToPublicKeyHash
+    {
+        if *height >= *n {
+            self.push(StackElement::Satisfied);
+            Some(Ok(SatisfiedConstraint::AbsoluteTimeLock { time: n }))
+        } else {
+            Some(Err(Error::AbsoluteLocktimeNotMet(n.clone())))
+        }
+    }
+
+    fn evaluate_sha256<'desc, Pk, Pkh>(
+        &mut self,
+        hash: &'desc sha256::Hash,
+    ) -> Option<Result<SatisfiedConstraint<'desc, 'stack, Pk, Pkh>, Error>>
+        where Pk: ToPublicKey, Pkh: ToPublicKeyHash
+    {
+        if let Some(StackElement::Push(preimage)) = self.pop() {
+            if preimage.len() != 32 {
+                return Some(Err(Error::CouldnotEvaluate("Hash preimage must be 32".to_owned())))
+            }
+            if sha256::Hash::hash(preimage) == *hash {
+                self.push(StackElement::Satisfied);
+                Some(Ok(SatisfiedConstraint::HashLock {
+                    hash: HashLockType::Sha256HashLock(hash),
+                    preimage,
+                }))
+            } else {
+                Some(Err(Error::CouldnotEvaluate("SHA2 hash failed".to_owned())))
```

Comment:
Please use multiple variants rather than strings. Strings undermine the type system and require extra allocations.